### PR TITLE
Add Monitor actor

### PR DESCRIPTION
**Pull request by K1li4nL**
This pr add the new actor for the scala backend: Monitor

This actor is tasked with the scheduling of heartbeats.

### CODE REVIEW

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      timers.cancelAll()
+
+    case Monitor.TriggerHeartbeat =>
+      log.info("triggering a heartbeat")
+      timers.cancel(singleHbKey)
+      heartbeatGenRef ! Monitor.GenerateAndSendHeartbeat(connectionMediatorRef)
+
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      jsonRpcMessage.getParams match {
+
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:49]: Is this line return intended (even if it is a small pattern matching)?

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  override def afterAll(): Unit = {
+    // Stops the test actor system
+    TestKit.shutdownActorSystem(system)
+  }
+
+  test("monitor should start scheduling heartbeats when told a server connected") {
+
+    val testProbe = TestProbe()
+    val monitorRef = system.actorOf(
+      Monitor.props(testProbe.ref, heartbeatRate = 2.seconds, messageDelay = 1.seconds)
```
Comment (Jharaxus) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:26]: These test settings variables (n seconds) are repeated across the tests so they could gain consistency by becoming attributes of the class

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    )
+
+    testProbe.send(monitorRef, Monitor.AtLeastOneServerConnected)
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat]
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat]
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat]
```
Comment (Jharaxus) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:32]: It may be more reliable to add our own timeout limit instead of relying on akka's default one

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  test("monitor should schedule single heartbeat when receiving a Right GraphMessage") {
```
Comment (Jharaxus) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:35]: This test also tests that the Right(message) only HB if message is a JsonRpcRequest without params with map, so this should be reflected in the test name or be put into a separate test (lines 49-50 and 55-56)

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    testProbe.expectNoMessage(5.seconds)
+
+    Source.single(Right(JsonRpcRequestExample.subscribeRpcRequest)).to(sink).run()
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat]
+
+    Source.single(Right(getMessagesById)).to(sink).run()
+    testProbe.expectNoMessage(5.seconds)
+  }
+
+  test("monitor should send heartbeats only when servers are connected") {
```
Comment (Jharaxus) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:59]: This test could also check that no HB is sent upon receiving a Right(JsonRpcRequest) that would otherwise expect a single HB

**Review by Jharaxus — CHANGES_REQUESTED** [review]
Minor changes, but the overall code is great!

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        case _: ParamsWithMap => /* Actively ignoring this specific message */
+        // For any other message, we schedule a single heartbeat to reduce messages propagation delay
+        case _ =>
+          if (!timers.isTimerActive(singleHbKey) && timers.isTimerActive(periodicHbKey)) {
```
[the remark this answers is not in the document]
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:53]: Definitely, fixed !

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala (hunk 1 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala]: Fixed

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala (hunk 4 above)
Comment (K1li4nL) [reply be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:39]: I decided to break it into two, couldn't find a good name that would express all that concisely.

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala (hunk 5 above)
Comment (K1li4nL) [reply be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:79]: Also done !

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala [hunk 7]
```diff
@@ -25,18 +25,23 @@ final case class Monitor(
...
   // Monitor is self-contained,
   // To that end it doesn't know the ref of the connectionMediator
   private var connectionMediatorRef = ActorRef.noSender
 
   override def receive: Receive = LoggingReceive {
 
     case Monitor.AtLeastOneServerConnected =>
+      someServerConnected = true
       connectionMediatorRef = sender()
       if (!timers.isTimerActive(periodicHbKey))
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:40]: You could use `someServerConnected` for this condition too

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 8]
```diff
@@ -23,58 +27,76 @@ class MonitorSuite extends TestKit(ActorSystem("MonitorSuiteActorSystem")) with
...
-    testProbe.expectNoMessage(5.seconds)
+    testProbe.expectNoMessage(timeout)
   }
 
   test("monitor should send heartbeats only when servers are connected") {
 
     val testProbe = TestProbe()
     val monitorRef = system.actorOf(
-      Monitor.props(testProbe.ref, heartbeatRate = 1.seconds, messageDelay = 60.seconds)
+      Monitor.props(testProbe.ref, heartbeatRate = fastRate, messageDelay = slowRate)
```
Comment (Jharaxus) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:83]: `messageDelay` should be `fastRate` otherwise even if an unexpected single HB is scheduled it would not be sent before the test timeout

**Review by Jharaxus — CHANGES_REQUESTED** [review]

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala (hunk 7 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:40]: Not sure, the logic here is that I want the timer to be set again only when it wasn't, regardless of server being connected or not. I just want to avoid the very narrow corner case where for some reason `Monitor` receives repetitively `AtLeastOneServerConnected` and have the timer reset every time.

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala (hunk 8 above)
Comment (K1li4nL) [reply be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:83]: Good point ! The test wasn't very insightful that way...

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala (hunk 7 above)
Comment (Jharaxus) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:40]: You can do the following to use `someServerConnected`:
```scala
connectionMediatorRef = sender()
if (!someServerConnected) {
  timers.startTimerWithFixedDelay(periodicHbKey, TriggerHeartbeat, heartbeatRate)
  someServerConnected = true
}
```

**Review by Jharaxus — DISMISSED** [review]

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    // Wait for the first heartbeat then "disconnect" servers
+    testProbe.expectMsgType[Monitor.GenerateAndSendHeartbeat](timeout)
+    testProbe.send(monitorRef, Monitor.NoServerConnected)
+    testProbe.expectNoMessage(timeout)
+
+    // No single heartbeat should be scheduled either
+    Source.single(Right(JsonRpcRequestExample.subscribeRpcRequest)).to(Monitor.sink(monitorRef)).run()
+    testProbe.expectNoMessage(timeout)
+
+    // Connect a server and check for heartbeats again
```
Comment (Ajkunas) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:98]: I do not think that those last lines do really belong to the corresponding test. 
You already tested it in previous tests, and it makes the test a bit confusing. I understand that the purpose was to test corner cases, but maybe writing another test by specifying in the description that you want to test if by connecting and disconnecting the servers, the Monitor still behaves correctly.

**Review by Ajkunas — CHANGES_REQUESTED** [review]
Overall really good ! 
I only have some minor comments that can be discussed. :)

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        case _: ParamsWithMap => /* Actively ignoring this specific message */
+        // For any other message, we schedule a single heartbeat to reduce messages propagation delay
+        case _ =>
+          if (someServerConnected && !timers.isTimerActive(singleHbKey)) {
+            log.info(s"Scheduling single heartbeat")
+            timers.startSingleTimer(singleHbKey, TriggerHeartbeat, messageDelay)
+          }
+      }
+
+    case _ => /* DO NOTHING */
```
[the remark this answers is not in the document]
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:64]: I am not sure I see the point, this last case is mostly filtering out Left(...) messages for which we don't want to trigger heartbeats. Messages received are already pretty well logged, I feel like this won't add much debugging value and will cluter the logs.

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+      timers.cancelAll()
+
+    case Monitor.TriggerHeartbeat =>
+      log.info("triggering a heartbeat")
+      timers.cancel(singleHbKey)
+      heartbeatGenRef ! Monitor.GenerateAndSendHeartbeat(connectionMediatorRef)
+
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      jsonRpcMessage.getParams match {
+        case _: ParamsWithMap => /* Actively ignoring this specific message */
```
[the remark this answers is not in the document]
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:55]: Again, I am worried about cluttering the logs, the server will potentially receive a lot of these so their logs might grow very fast

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala (hunk 9 above)
Comment (K1li4nL) [reply be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:98]: The test goes as follows: 

`monitor should send heartbeats only when servers are connected`

1. Initiate monitor by telling it about connected servers
2. expect a heartbeat
3. disconnect the servers
4. expect no heartbeat
5. connect servers again 
6. expect heartbeats

Excluding these last steps 5 and 6 make the test look incomplete to me (what if monitor has a bug that prevent it to start again after some servers came back). I want the test to guarantee that monitor does its job _whenever and only when servers are connected_.  And it looks redundant to have a second test just doing that last bit.

Do you think the name : 
`monitor should send heartbeats if and only if servers are connected` 

Would make it clearer ?

Comment (Ajkunas) [reply be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:98]: Okay I see your point and I agree that the second test would be redundant!
The new name could be a bit clearer in my opinion. :)

**Review by Ajkunas — APPROVED** [review]
